community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java

```diff
@@ -0,0 +1,549 @@
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.neo4j.collection.primitive.PrimitiveLongSet;
+import org.neo4j.helpers.collection.Visitor;
+import org.neo4j.kernel.api.exceptions.index.FlipFailedKernelException;
+import org.neo4j.kernel.api.exceptions.index.IndexCapacityExceededException;
+import org.neo4j.kernel.api.exceptions.index.IndexPopulationFailedKernelException;
+import org.neo4j.kernel.api.index.IndexConfiguration;
+import org.neo4j.kernel.api.index.IndexDescriptor;
+import org.neo4j.kernel.api.index.IndexEntryConflictException;
+import org.neo4j.kernel.api.index.IndexPopulator;
+import org.neo4j.kernel.api.index.IndexUpdater;
+import org.neo4j.kernel.api.index.NodePropertyUpdate;
+import org.neo4j.kernel.api.index.PropertyAccessor;
+import org.neo4j.kernel.api.index.Reservation;
+import org.neo4j.kernel.api.index.SchemaIndexProvider;
+import org.neo4j.kernel.api.index.SchemaIndexProvider.Descriptor;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+import org.neo4j.register.Register.DoubleLong.Out;
+import org.neo4j.register.Register.DoubleLongRegister;
+import org.neo4j.register.Registers;
+
+import static java.lang.String.format;
+import static org.neo4j.kernel.impl.api.index.IndexPopulationFailure.failure;
+
+/**
+ * {@link IndexPopulator} that allow population of multiple indexes during one iteration.
+ * Performs operations by calling corresponding operations of particular index populators.
+ */
+public class MultipleIndexPopulator implements IndexPopulator
+{
+    private final Queue<NodePropertyUpdate> queue = new ConcurrentLinkedQueue<>();
+    private final IndexStoreView storeView;
+    private final LogProvider logProvider;
+    private final Log log;
+    private final List<Populator> populators = new ArrayList<>();
+
+    public MultipleIndexPopulator( IndexStoreView storeView, LogProvider logProvider )
+    {
+        this.storeView = storeView;
+        this.logProvider = logProvider;
+        this.log = logProvider.getLog( IndexPopulationJob.class );
+    }
+
+    public void addPopulator(
+            IndexPopulator populator,
+            IndexDescriptor descriptor,
+            Descriptor providerDescriptor, IndexConfiguration config,
+            FlippableIndexProxy flipper,
+            FailedIndexProxyFactory failedIndexProxyFactory,
+            String indexUserDescription )
+    {
+        populators.add( new Populator( populator, descriptor, config, providerDescriptor,
+                flipper, failedIndexProxyFactory, indexUserDescription ) );
+    }
+
+    @Override
+    public void create()
+    {
+        for ( int i = 0; i < populators.size(); )
+        {
+            Populator populator = populators.get( i );
+            log.info( "Index population started: [%s]", populator.getIndexUserDescription() );
+            try
+            {
+                populator.create();
+                i++;
```

Comment:
I think it is a bit confusing that you remove things behind the scenes from the populator list, but I am not gonna fight against it. Just a couple of things: 1) can we add a comment explaining this? and 2) can we use a linked-list instead of an array list to avoid array copies?